core/data-layer/src/data_layer.rs

```diff
@@ -696,6 +731,121 @@ impl DataLayer {
     }
 
     // Source Metrics
+    pub async fn add_source_metrics_in_data_layer(
+        &self,
+        _collector: &str,
+        metric_id: &str,
+        json_value: &str,
+    ) -> Result<()> {
+        /* [ Comment ]
+         *  source metrics: Metric data is separated by metric_id and ulid is sorted in ascending order (using for ScalingPlan search)
+         *  source metrics metadata: Metric data is sorted in ASC order by ULID (using for remove target data to maintain buffer size)
+         * [ Data structure ]
+         *  source metrics - HashMap<key: metric_id, value: BTreeMap<key: ULID, value: SourceMetrics>>
+         *  source metrics metadata - LinkedList<(metric_id, ULID, data size(source metrics + source metrics metadata)> - list order by ULID ASC */
+        let mut source_metrics = self.source_metrics_data.source_metrics.write().await;
+        let mut source_metrics_metadata = self
+            .source_metrics_data
+            .source_metrics_metadata
+            .write()
+            .await;
+        let mut source_metrics_size = self.source_metrics_data.source_metrics_size.write().await;
+        let metric_buffer_size_kb = self.source_metrics_data.metric_buffer_size_kb;
+
+        let now_ulid = Ulid::new().to_string();
+        let source_metric_insert_data = SourceMetrics {
+            json_value: json_value.to_string(),
+        };
+        // source_metric_insert_data_size = source metrics + source metrics metadata
+        let source_metric_insert_data_size = source_metric_insert_data.get_heap_size()
+            + (metric_id.to_string().get_heap_size() * 2)
+            + (now_ulid.get_heap_size() * 2);
+
+        let mut total_source_metrics_size = *source_metrics_size + source_metric_insert_data_size;
+
+        // get remove target data
+        let mut remove_target_data: Vec<(String, String, usize)> = Vec::new();
+        let mut subtract_total_size = total_source_metrics_size;
+        loop {
+            // check size :: buffersize - total size < 0 (None) => remove target data
+            if metric_buffer_size_kb
+                .checked_sub(subtract_total_size as u64)
+                .is_none()
+            {
+                let Some(front_source_metrics_metadata) = source_metrics_metadata.pop_front() else {
+                        break;
+                    };
+                subtract_total_size -= front_source_metrics_metadata.2; // oldest metadata size
+                remove_target_data.append(&mut vec![front_source_metrics_metadata]);
+                continue;
+            }
+            break;
+        }
+
+        // remove target data
+        remove_target_data
+            .iter()
+            .for_each(|(metric_id, ulid, size)| {
+                let Some(metric_id_data) = source_metrics.get_mut(metric_id) else {
```

Comment:
```suggestion
    let Some(source_metrics_map) = source_metrics.get_mut(metric_id) else {
```

to match variable names as shown below (source_metrics_map)